### DISCUSSION TITLE
Limits for the SAR in a '.seq'  file for Siemens systems

### DISCUSSION THREAD

**Discussion by ZemaTimoteo** [pulseq/pulseq#59] [Q&A] [answered]
Hi all,

First of all, thank you for the great work you did with Pulseq. 

I encounter an issue regarding the SAR when using a '.seq' Multi-echo Spin-Echo (MESE) sequence in a 3T Siemens system, with some slices (i.e. Nslices=18) - (I can share the '.seq' file if you want/need)

In short, I designed this sequence in a way that it had a B1rms <10uT. 
a) In the first design, with 20 slices, when I tried to run it in the scanner, I received an alert telling me that the system couldn't run it. 
b) In the second design, I increased a bit the TR and reduced the number of slices, as I was loading the sequence, I did not receive the same alert, however, after 5-6s the system blocked the sequence box with a sentence in red saying 'Can not run any sequence'. Then I was seeing an alert from the system telling me to restart it, or shut it down or call for Siemens support.

QUESTION: 
a) How to deal with SAR in a Siemens System? 
b) What are the safety limits that I should take into consideration?
c) How can I measure them in my '.seq' file?

Thank you in advance for the Help
Feel free to ask me more  about this

Best
Tiago

**Comment by QingpingChen1201** [comment, marked as answer]
Hi Tiago,
Many thanks for your interest in Pulseq!
## Answer to Question a): How to deal with SAR in a Siemens System?
**_MR safety_** is always the first and the most important thing we should take into consideration when designing a sequence (no matter in Pulseq or not). MR sequences all use electromagnetic radiofrequency (RF) pulses. RF electromagnetic waves are used to generate signal in the patient. Unlike x-rays, for example, the RF used in MRI is not ionizing. However, the RF used by MRI scanners causes some heating by depositing energy into the tissue. **_Note that this heating is patient, scanner, and sequence dependent._** The safety of these pulses can be assessed by using either SAR or B1+ RMS: 
* SAR stands for Specific Absorption Rate and describes the rate of energy deposition into tissue.
* B1+ is used to describe the amplitude of the RF magnetic field. B1+ RMS is the average effective magnetic field generated by a particular transmit coil for a particular sequence.         

The Siemens system (or other vendors' system) **estimates** and **restricts** SAR according to the operating mode of the machine which is set to match limits specified by the International Electrotechnical Commission (IEC). There are different limits relating to different parts of the body (the limits are more strict for implants). These limits practically implement more general safety guidance that restricts the allowed whole-body temperature rise to less than 1 Celsius degree.     
When you set up an exam and load a sequence, you should be able to check the estimated SAR and B1+ RMS provided by the system. When your sequence provides an estimated SAR that violates the limit set by the system (for a specific coil and specific patient), you won't be able to run your sequence. In this case, you should **_loose_** your sequence a bit to satisfy the SAR requirements or switch to a less-strict operating mode (e.g. First-Level mode).

## Answer to Question b): What are the safety limits that I should take into consideration?
As we mentioned above, SAR is dependent on sequence, scanner, and patient. We cannot change the scanner, transmit coil, and patients. But we can adjust our sequences to satisfy the safety limits. The sequence-specific parameters we need to take into consideration include TR, RF amplitude, RF duration, the duration between two RF pulses, etc.

## Answer to Question c): How can I measure them in my '.seq' file?
There are two functions for SAR calculation in Pulseq: `mr.calcRfPower(rf, dt)` for RF pulses, and `seq.calcRfPower()` for the whole sequence. These two functions are in the current "dev" branch of the Pulseq Github repository: https://github.com/pulseq/pulseq/blob/dev/matlab/%2Bmr/calcRfPower.m  and https://github.com/pulseq/pulseq/blob/dev/matlab/%2Bmr/%40Sequence/Sequence.m.
With these two functions, it is possible to calculate a relative SAR. However, It is not possible for Pulseq to calculate the correct absolute SAR. SAR values can vary significantly between scanners and patients. The absolute SAR is related to the electric field, so the further scaling coefficient is both transmit-coil-dependent (e.g. depends on the coil design) and also subject-dependent.
You can use this line: `[total_energy, peak_pwr, rf_rms] = seq.calcRfPower()` to calculate the relative total energy, RF power, and B1 RMS for your Pulseq sequence.
we compared the predictions provided by the `seq.calcRfPower()` function with the SAR / B1rms values estimated by the scanner and have received an excellent agreement with regard to B1 RMS and a very good linear correlation to SAR. 

Please feel free to let us know if you need any further information.

Best regards,
Qingping and the Pulseq team

**Reply by ZemaTimoteo** [reply]
Thank you so much for the complete answer.

I have a follow up question:

What are the (numerically) limits that you consider for rf_rms or to total energy , regarding the Siemens vendor equipment? In other words, within which values do you feel confident that the sequence will run in a Siemens equipment?

Best regards
Tiago T Fernandes

**Reply by QingpingChen1201** [reply]
Hi Tiago,
many thanks for the question. It is a bit difficult for me to say the range of the B1 RMS or total energy within which the sequence can run in a Siemens equipment. This is because different scanners, coils, and patients have different requirement on the B1 RMS or SAR. I would say for the sequences that are not really "heavy" (e.g., not RF-intense or the TR is not really short), we don't need to worry about the SAR limits. But heavy sequences, such as the multi spin echo sequence, are much more likely to violate the SAR limits. In this case, you can adjust the sequence several times to hit the SAR limits required by your scanner, coil, and patient. And then you can roughly get to know the absolute values (B1 RMS or total energy) required for the specific use case.
Please feel free to let us know if you have any further questions or need any further information.
Best regards,
Qingping and the Pulseq team

**Comment by ZemaTimoteo** [comment]
Hi Qingping, 

I have a further question related to the rf_rms. 
I adapted the MATLAB code on the github (_see Figure 2_) to Python:
![figure2](https://github.com/user-attachments/assets/29eb2931-2130-4a55-8e06-f9dd35d39099)

```

# Pulseq Implementation from github Pulseq Matlab
dt          = 1e-06   # (in s)
rf_rms    = 0
n_slices  = 25
n_phA    = 120      # Ny

# Resample the pulse to a reasonable time array
nn_ex  = round(t_ex/dt)
t          = ( np.linspace(0,(nn_ex-1),nn_ex) + 0.5 ) * dt    # in (s)

rfs       = np.interp(t,rf_ex.t,rf_ex.signal)                       
rfs_sq  = np.multiply(rfs,np.conj(rfs))

total_energy   = sum(rfs_sq)*dt                                       
peak_pwr        = max(rfs_sq)
aux_rf_rms_ex = np.sqrt(total_energy/t_ex)                     

# rf_rms = ( aux_rf_rms**2 * t_ex ) / gamma *  n_slices * n_phA  # T
#rf_rms =( ( aux_rf_rms_ex**2 * t_ex ) / gamma)  * 1e6 # uT
rf_rms = ( ( aux_rf_rms_ex**2 * t_ex ) )  # uT s

nn_ref     = round(t_ref / dt)
t_intr_ref = (np.linspace(0, (nn_ref - 1), nn_ref) + 0.5) * dt

for k_echo in range(n_echo):  # cicle per TR each with one phase (line of k_space) and multiple Echos
    rfs_ref      = np.interp(t_intr_ref, rf_ref[k_echo].t, rf_ref[k_echo].signal)                          # (in Hz)
    rfs_ref_sq = np.multiply(rfs_ref, np.conj(rfs_ref))

    total_energy_ref = sum(rfs_ref_sq) * dt   # uT s
    total_energy       = total_energy + total_energy_ref
    peak_pwr_ref      = max(rfs_ref_sq)
    aux_rf_rms_ref    = np.sqrt(total_energy_ref / t_ref)


    aux_rf_rms =  ( ( aux_rf_rms_ref ** 2 * t_ref) )   # uT s
    rf_rms        = rf_rms + aux_rf_rms                   # uT s

# B1_rms+ for one ETL
aux_rf_rms = np.sqrt(rf_rms / TE_train)   # Hz

# Get for all slices and real Ny
B1_rms = aux_rf_rms  *  n_slices * n_phA  / gamma  *1e6  # uT s (gamma - Hz/Tesla)
```

 and run it on the .seq file that I attach. 
test523_MSE_vFA_T2-45ms_nEchos-13_nSlices-25_TE-9ms_TR-4133ms_FOV-160mm_Nx-320_Ny-320_gm-32_sm-120_sTexc-3mm_sTrefoc-9mm_acc_LORAKS.zip)


On IDEA, I obtained for the same n_slices =25 and n_phA = Ny = 120, the following values, _see figure 1_. ![figure1](https://github.com/user-attachments/assets/5edab2f2-4338-49a6-91e7-d2490f675d65)

On the Python implementation, I obtained: 

- 17138 uT


Regarding this, I need support to understand the reason why I obtained such different values between the pulseq implementation and the IDEA, when you stated that they have a very good similarity.

Since I am working with an MSE sequence, I must understand the B1+rms. Could you help me figure this out to gain trust in the measurement of rf_rms with the pulseq tools.

**Comment by QingpingChen1201** [comment]
Hi Tiago,
Thanks for your updates. I have little idea about the simulated B1_Max values in the IDEA virtual machine. We can obtain the vendor-provided B1+ RMS value directly from the scanner console: you can run the sequence and then click SAR at the bottom to see the SAR-related values, please see the figure below.
![SAR_UI](https://github.com/user-attachments/assets/4cb0a8d9-5970-49ce-a94c-b735a9169b54)
For example, for a very long EPI sequence, the Pulseq-calculated B1+ RMS is 42.0423 Hz = 0.9874 uT, and the vendor-provided B1+ RMS is 0.96 uT. For a single spin-echo sequence, the Pulseq-calculated B1+ RMS is 89.6285 Hz = 2.1049 uT, and the vendor-provided B1+ RMS shown in the console is 1.99 uT. The Pulseq-calculated B1+ RMS values agree quite well with the vendor-provided values.
Hope this helps!
Best regards,
Qingping and the Pulseq team

**Reply by ZemaTimoteo** [reply]
Dear Qingping,

Thank you for your valuable feedback. In a subsequent iteration of this issue, I encountered additional questions that I would like to discuss.

Firstly, I was able to replicate the case for the EPI sequence. However, I was unable to locate the SE sequence design Matlab file.
- Would it be possible for you to provide it?

Secondly, while attempting to implement my MSE sequence, I observed a discrepancy between the B1+ RMS values provided by the vendor and those calculated by Pulseq. Upon further investigation, I hypothesized that this discrepancy could be related to the timing of each RF pulse. Specifically, I found that when the time between rf_exc and rf_refoc was similar (i.e. t=1.8ms), the B1+ RMS values were also comparable (B1+rms vendor = 0.22uT vs B1+rms pulseq = 0.221uT). However, when these times differed, the scanner tended to calculate a lower B1+ RMS value.

My questions are as follows:

- Have you encountered similar behavior in your experiments?
- Do you know if the duration of the RF pulse affects its real amplitude when compiled by the scanner?

Finally, I also experienced in the past troubles with the Duty cycle, do you have any experience with the scanner in order to have a report of its intensity?  

I appreciate your insights on these matters. If any doubt please reach out.

Best regards,

**Reply by QingpingChen1201** [reply]
Hi Tiago,
The EPI and SE sequences are the ones I used for a quality assurance paper submitted to MAGMA. The paper will come out soon, you are welcome to read it. You can access the source code from [here](https://github.com/HarmonizedMRI/qualityAssurance/tree/main/data_acquisition).
For the second and third questions, please refer to Maxim's answer above.
Regarding "duty cycle", you mean the percent time in a pulse sequence RF is being transmitted? I am not aware of the report of its intensity provided by the scanner. SAR limits restrict RF heating effects. The SAR increases with field strength, RF power and duty cycle, type of the transmitter coil, and body size. In high and ultrahigh magnetic fields (e.g. 3T), important SAR issues can be readily addressed by, e.g., reducing the RF duty cycle with longer TR and the use of parallel imaging techniques.
For more detailed information about SAR, I recommend you to read this article [here](https://mriquestions.com/what-is-sar.html).
Please feel free to let us know if you need any further information.
Best regards,
Qingping and the Pulseq team

**Comment by m-a-x-i-m-z** [comment]
Dear Tiago, 

the only instance when we've observed deviations between the Pulseq's B1 RMS estimation and the one from the scanner was when the RF pulse amplitudes were trimmed by the scanner. Please always select "Timing and Flip Angles" -> Strict on the Sequence -> Special card if you really want accurate flip angles. Be advised, that this is not compatible with table motion, e.g. if the scanner moves the table before the scan to move the imaging volume to the physical iso-center, the scan is likely to fail. 

Best regards, 
M.

**Reply by ZemaTimoteo** [reply]
Dear Maxim,

Thank you for your input. With further tests, I still cannot match the b1+rms calculated with pulseq with the one measured on Siemens, for my MSE-implemented sequence. Probably the scanner is applying some transformation to it. 
So I have a couple of further questions:
- What do you mean by 'trimmed'? What sort of transformation does this process do?
- I can only find the option 'Strict' on a GRE sequence. In the MSE sequence with variable FA, I cannot find such an option. Do you have any explanation for that? (both sequences were compiled using pypulseq v1.3.1)
- Any other way to try to mitigate this effect? 

Thank you so much in advance,

Best
Tiago

**Comment by m-a-x-i-m-z** [comment]
I see, this was exactly my point: if you cannot see the option "strict" for your sequence it means that the flip angles are not realizable and are readily "trimmed". Try making pulses in your sequence longer to reduce the max B1 and you should be able to select "strict". If "strict" is not available there always will be  a mismatch between the predicted and applied B1_rms because the scanner scales down some of the pulses. 

Best,
Maxim & the Pulseq team

**Comment by ZemaTimoteo** [comment]
Thank you for your preciosous input Maxim

I was able to both achieved a 'strict' option on the scanner, as well as matching values of b1+rms between measured and scanner. However, I still encountered an issue when exploring the impact of b1+rms.

I encountered a mismatch in terms of measurement of the slew rate between the pulseq and the SIEMENS. 
My MSE sequence has the particularity of having different times for the t_ex and t_ref and a slice thickness which is settled to be 3x higher in the refocusing than in the excitation (according to literature it improves the slice profile). When I tried to run this in the scanner it failed to run. When I tested in the IDEA simulator, it did not run as well. In the err file it had a slew rate of the gz higher than the one predicted in pulseq and over the limits of the machine. 

When checked on pulseq I had the following values of slew rate and maxAmplitude of gz_ex and gz_ref (see Figure) for a ST_ex = 3mm, ST_ref = 9mm, t_ex2.7ms, t_ref1.25ms:
![Captura de ecrã 2025-02-07 171743](https://github.com/user-attachments/assets/5cc97f24-88d4-4d18-a7a1-aee943b5533e)


When I had a mismatch between maxAmplitude or slew rate between pulses I would have a failed test in IDEA, however, if they were similar it would work out. Why pulseq does not calculate the slew rate and amplitude the same way as IDEA in such cases?

**Reply by m-a-x-i-m-z** [reply]
did you try running seq.testReport on you sequence? you may find examples how to do it at the end o many demo sequences.

**Reply by ZemaTimoteo** [reply]
I did double checked and, for this '.seq' file (in the .zip file), I got the following discrepancy (Err file from IDEA vs test_report from pypulseq==1.3.1.post1):

![imagem](https://github.com/user-attachments/assets/ee0cde53-a2a2-4ce6-b540-ac03ed09f1fd)
[test592_MSE_vFA_T2-45ms_nEchos-1_nSlices-1_TE-8ms_TR-4133ms_FOV-256mm_Nx-262_Ny-1_gm-32_sm-120_sTexc-3mm_sTrefoc-9mm_acc_noAcc.zip](https://github.com/user-attachments/files/18717953/test592_MSE_vFA_T2-45ms_nEchos-1_nSlices-1_TE-8ms_TR-4133ms_FOV-256mm_Nx-262_Ny-1_gm-32_sm-120_sTexc-3mm_sTrefoc-9mm_acc_noAcc.zip)

**Reply by QingpingChen1201** [reply]
this sounds strange to me. I cannot load your .seq file because the Pulseq version (v1.3.1) is a bit low. Maybe you can try to update your Pulseq version a bit to the latest version and try again. The new Pulseq v1.5.0 version will be out soon.
Best regards,
Qingping

**Comment by m-a-x-i-m-z** [comment]
Hi! 

can it be that you were running your sequence with the "adapt timing" option? Why are you then surprised that something is not correct? Adaptive timing should only be used for the simplest demo sequences, not for anything serious. Please always use the strict timing mode in the future. 

Secondly, you seem to be using ancient Pulseq libraries which generate you the Pulseq file in strictly speaking not a standard conforming way, e.g. the version header is not only based on numbers but contains text characters, but this is a minor problem. 

The actual problem is related to your RF pulse in block 4, which only has a delay of 70us, whereas the system typical RF dead time is 100 us. This is why your sequence simply cannot run correctly in the strict timing mode. 

Little bit off-topic, but your ADC dwell time is also not compatible which what the system can realize. At least seq.checkTiming() picks this up for me. 

Hope this helps!

Best,
Maxim
